src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+    }
+
+    GenTree* t1Comp = t1.compTree;
+    t1Comp->SetOper(cmpOp);
+    t1Comp->AsOp()->gtOp1         = cmpOp1;
+    t1Comp->AsOp()->gtOp2->gtType = foldType; // Could have been varTypeIsGC()
+    if (optReturnBlock)
+    {
+        // Update tree when b1: BBJ_COND and b2: GT_RETURN (BBJ_RETURN)
+        t1Comp->AsOp()->gtOp2->AsIntCon()->gtIconVal = 0;
+        t1.testTree->gtOper                          = GT_RETURN;
+        t1.testTree->gtType                          = t2.testTree->gtType;
+    }
+
+#if FEATURE_SET_FLAGS
+    // For comparisons against zero we will have the GTF_SET_FLAGS set
+    // and this can cause an assert to fire in fgMoveOpsLeft(GenTree* tree)
+    // during the CSE phase.
+    //
+    // So make sure to clear any GTF_SET_FLAGS bit on these operations
+    // as they are no longer feeding directly into a comparisons against zero
+
+    // Make sure that the GTF_SET_FLAGS bit is cleared.
+    // Fix 388436 ARM JitStress WP7
+    c1->gtFlags &= ~GTF_SET_FLAGS;
+    c2->gtFlags &= ~GTF_SET_FLAGS;
+
+    // The new top level node that we just created does feed directly into
+    // a comparison against zero, so set the GTF_SET_FLAGS bit so that
+    // we generate an instruction that sets the flags, which allows us
+    // to omit the cmp with zero instruction.
+
+    // Request that the codegen for cmpOp1 sets the condition flags
+    // when it generates the code for cmpOp1.
+    //
+    cmpOp1->gtRequestSetFlags();
+#endif
+
+    if (!optReturnBlock)
+    {
+        // Update edges if b1: BBJ_COND and b2: BBJ_COND
+
+        flowList* edge1 = comp->fgGetPredForBlock(b1->bbJumpDest, b1);
+        flowList* edge2;
+
+        if (sameTarget)
+        {
+            edge2 = comp->fgGetPredForBlock(b2->bbJumpDest, b2);
+        }
+        else
+        {
+            edge2 = comp->fgGetPredForBlock(b2->bbNext, b2);
+
+            comp->fgRemoveRefPred(b1->bbJumpDest, b1);
+
+            b1->bbJumpDest = b2->bbJumpDest;
+
+            comp->fgAddRefPred(b2->bbJumpDest, b1);
+        }
+
+        noway_assert(edge1 != nullptr);
+        noway_assert(edge2 != nullptr);
+
+        BasicBlock::weight_t edgeSumMin = edge1->edgeWeightMin() + edge2->edgeWeightMin();
+        BasicBlock::weight_t edgeSumMax = edge1->edgeWeightMax() + edge2->edgeWeightMax();
+        if ((edgeSumMax >= edge1->edgeWeightMax()) && (edgeSumMax >= edge2->edgeWeightMax()))
+        {
+            edge1->setEdgeWeights(edgeSumMin, edgeSumMax, b1->bbJumpDest);
+        }
+        else
+        {
+            edge1->setEdgeWeights(BB_ZERO_WEIGHT, BB_MAX_WEIGHT, b1->bbJumpDest);
+        }
+    }
+
+    /* Modify the target of the conditional jump and update bbRefs and bbPreds */
+
+    if (optReturnBlock)
+    {
+        comp->fgRemoveRefPred(b1->bbJumpDest, b1);
+
+        noway_assert(!b2->bbJumpDest);
+        b1->bbJumpDest = b2->bbJumpDest;
+        b1->bbJumpKind = b2->bbJumpKind;
+        b1->bbJumpSwt  = b2->bbJumpSwt;
```

Comment:
```suggestion
        b1->bbJumpKind = BBJ_RETURN;
```
You shouldn't need to touch bbJumpSwt. Maybe nulling out bbJumpDest makes sense? Or maybe only needed in debug?

Reply:
Done.
```
        m_b1->bbJumpDest = nullptr;
        m_b1->bbJumpKind = BBJ_RETURN;
#ifdef DEBUG
        m_b1->bbJumpSwt  = m_b2->bbJumpSwt;
#endif
```